File: Medium/1052 - GrumpyBookstoreOwner/solution.ts
// Question 1052. Grumpy Bookstore Owner (https://leetcode-cn.com/problems/grumpy-bookstore-owner/)

/**
 * 执行用时：88 ms, 在所有 TypeScript 提交中击败了 100.00% 的用户
 * 内存消耗：42.3 MB, 在所有 TypeScript 提交中击败了 100.00% 的用户
 *
 * @param {number[]} customers
 * @param {number[]} grumpy
 * @param {number} X
 * @returns {number}
 */
function maxSatisfied(
  customers: number[],
  grumpy: number[],
  X: number
): number {
  const MINUTES: number = customers.length;
  let total: number = 0;
  for (let i = 0; i < MINUTES; i++) {
    total += (1 - grumpy[i]) * customers[i];
  }

  let increase: number = 0;
  let maxIncrease: number = 0;

  for (let i = 0; i < X; i++) {
    increase += grumpy[i] * customers[i];
  }
  maxIncrease = increase;

  for (let i = X; i < MINUTES; i++) {
    increase =
      increase - grumpy[i - X] * customers[i - X] + grumpy[i] * customers[i];
    maxIncrease = Math.max(maxIncrease, increase);
  }
  maxIncrease = Math.max(maxIncrease, increase);

  return total + maxIncrease;
}

File: Medium/1052 - GrumpyBookstoreOwner/Solution.cs
// Question 1052. Grumpy Bookstore Owner (https://leetcode-cn.com/problems/grumpy-bookstore-owner/)

/// <summary>
/// 执行用时：156 ms, 在所有 C# 提交中击败了 100.00% 的用户
/// 内存消耗：33.5 MB, 在所有 C# 提交中击败了 7.14% 的用户
/// </summary>
public class Solution 
{
    public int MaxSatisfied(int[] customers, int[] grumpy, int X) 
    {
        int MINUTES = customers.Count();
        int total = 0;
        for (var i = 0; i < MINUTES; i++)
        {
            total += (1 - grumpy[i]) * customers[i];
        }

        int maxIncrease = 0;
        int increase = 0;

        for (var i = 0; i < X; i++)
        {
            increase += grumpy[i] * customers[i];
        }
        maxIncrease = increase;

        for (var i = X; i < MINUTES; i++)
        {           
            increase = increase - grumpy[i - X] * customers[i - X] + grumpy[i] * customers[i];
            maxIncrease = Math.Max(maxIncrease, increase);
        }

        maxIncrease = Math.Max(maxIncrease, increase);

        return maxIncrease + total;
    }
}